compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/QueryDesugar.java

```diff
@@ -2499,7 +2500,13 @@ public void visit(BLangMatchStatement matchStmt) {
     public void visit(BLangMatchClause matchClause) {
         matchClause.matchPatterns.forEach(this::acceptNode);
         this.acceptNode(matchClause.matchGuard);
+        matchClause.declaredVars.forEach((k, v) -> identifiers.putIfAbsent(k, v));
```

Comment:
Even though shadowing is disallowed; let's not do this. Instead, save the initial `identifiers` and revert back to them as you exit the visit.
Also, you should **not** add only if it's a new symbol. Do `putAll` instead

Reply:
Updated in [#bf1e4645](https://github.com/ballerina-platform/ballerina-lang/pull/41045/commits/bf1e4645085535ef06fcffc22a51274788e43739).